### DISCUSSION TITLE
CLI should remove consumer keys in connected app meta files

### DISCUSSION THREAD

**Discussion by mcarvin8** [forcedotcom/cli#3038] [General]
https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_connectedapp.htm

In order to deploy connected apps using the Salesforce CLI, you have to delete the consumer key line manually or else the deployment will fail because the key is unique in each org until you do a refresh. 

I had to write a custom script to automatically delete the consumer key lines in every consumer app meta file before starting a deployment with connected apps. This works to deploy apps to multiple orgs with the other settings the same (previously I just had to instruct our developers to manually delete the line after retrieval). 

The Salesforce CLI should be updated to not include the consumer key when retrieving connected apps from an org so the files will be deployable automatically after retrieval.

**Comment by VivekMChawla** [comment]
@mcarvin8 - Thanks for posting this suggestion.  I have a couple of questions and a suggestion.

1. Is it conceivable that some users would want to **retrieve** the consumer key?
2. Would it be better to ignore the consumer key on **deployment**, instead of not including it when retrieving connected apps from an org?
3. Have you considered using the [string replacement](https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_ws_string_replace.htm) capabilities of the CLI to strip the `consumerKey` XML node on deployment? (thanks @WillieRuemmele for this suggestion!)

**Reply by mcarvin8** [reply]
@VivekMChawla - If the deployment command at least can ignore the consumer keys, I'm fine with leaving them in there. 

It just fails a deployment when left intact in the current condition. 

I in fact tried the string replacements and use them all the time for the bot user field in bots. 

This replacement does not work when trying to delete a line in the XML. And trying to clear the line in the xml (or just delete contents of the consumer key but leave the key element intact) doesn't work either with string replacements. 

If there's a string replacement that works, I'm fine with using. But you would need to tell everyone deploying connected apps to add that line to their sfdx-project.json file. Updating the deploy command to ignore the keys seems more accessible for all users.

**Reply by VivekMChawla** [reply]
@mcarvin8 - Thanks for the follow-up.  I see your point about the string replacement. It would be nice if this "just worked" for everyone, especially if there is **never** a valid use case for deploying a consumer key.

I'll raise this with the team and the Connected App PM to ensure there aren't unintended consequences from this suggested change and report back once I've heard their perspective.

**Reply by mcarvin8** [reply]
Thanks. Yup let me know if you hear any concerns regarding stripping consumer keys automatically during deployments. 

I haven't seen issues (nor anyone report issues to me with connected apps in multiple orgs) since implementing our custom script to do so (or since just telling developers to delete the line manually), but the connected app team would know better than me.

**Reply by pkozuchowski** [reply]
You can just remove consumerKey line from file in your repository and you can deploy that app everywhere. At least that's what I'm doing right now.

**Reply by mcarvin8** [reply]
Thanks @pkozuchowski . Yes, that's what we're doing too with a script. Hopefully Salesforce removes them or at least ignores them with the Salesforce CLI to automate this for all users.

**Reply by daveespo** [reply]
I think @pkozuchowski 's suggestion of removing the Consumer Key XML element from your project/repo is the right approach here. 

Having the CLI do black magic and omit Consumer Key on deployment would give a casual observer the impression that the Consumer Key in your repo is the one that is in your scratch org. That's not a great surprise to stumble upon after a few hours of troubleshooting. 

I acknowledge that a `retrieve` would end up restoring the Consumer Key in your local project so it just requires vigilance to review what you're planning to commit.

**Reply by aheber** [reply]
As far as I can remember pulling this file is the only way to retrieve that value.

In the past I have needed to know that value so I can use that new connected app to communicate with the org.

I wouldn't want to lose the ability to access that value, I would be OK if that behavior were managed with an environment variable. I just don't want it to become inaccessible.

**Reply by pkozuchowski** [reply]
@aheber You can check consumer key and secret from the UI.

**Reply by aheber** [reply]
I'm sorry I wasn't clear, I'm talking about a headless & automated environment. The API doesn't provide another mechanism.

**Reply by pkozuchowski** [reply]
@aheber  Oh yes, that's actually very good topic! Commands for headless rotation of secrets and certificates would be great
